Add codespell support (config, workflow to detect/not fix) and make it fix some typos

yarikoptic opened the pull request:
More about codespell: https://github.com/codespell-project/codespell .

I personally introduced it to dozens if not hundreds of projects already and so far only positive feedback.

CI workflow has 'permissions' set only to 'read' so also should be safe.

Let me know if just want to take typo fixes in and get rid of the CI

yarikoptic committed:
Add github action to codespell main on push and PRs

Signed-off-by: Yaroslav O. Halchenko <debian@onerussian.com>

yarikoptic committed:
Add rudimentary codespell config

Signed-off-by: Yaroslav O. Halchenko <debian@onerussian.com>

yarikoptic committed:
[DATALAD RUNCMD] chore: run codespell throughout fixing a few  new typos automagically

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "codespell -w",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

Signed-off-by: Yaroslav O. Halchenko <debian@onerussian.com>

yarikoptic commented:
I have read the CLA Document and I hereby sign the CLA

bolinfest reviewed (approved):
Nice! All the fixes are spot-on, so let's try it with the CI.

bolinfest commented on .github/workflows/codespell.yml:
```diff
@@ -0,0 +1,25 @@
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Annotate locations with typos
+        uses: codespell-project/codespell-problem-matcher@v1
```
Hmm, this does not appear to have seen much interest or activity:

https://github.com/codespell-project/codespell-problem-matcher

yarikoptic replied:
Yep, but it seems just to work in my experience

bolinfest replied:
Right, I trust that it works today. What I have less confidence in is a bad actor coming in, somehow gaining control of `codespell-problem-matcher`, changing the `v1` tag to point at something else, and anyone noticing.

yarikoptic replied:
yeap, there was a similar concern/practice elsewhere recently (https://github.com/argoproj/argo-workflows/pull/14454) -- I  will change now to current commit hashes...  ran `pin-github-action .github/workflows/codespell.yml` and pushed

bolinfest commented on .github/workflows/codespell.yml:
```diff
@@ -0,0 +1,25 @@
+      - name: Annotate locations with typos
+        uses: codespell-project/codespell-problem-matcher@v1
+      - name: Codespell
+        uses: codespell-project/actions-codespell@v2
```
This one is a bit better:

https://github.com/codespell-project/codespell-problem-matcher

yarikoptic replied:
Sorry, what do you mean, since it is the one listed above

bolinfest replied:
Copy/paste error: that should have been https://github.com/codespell-project/actions-codespell. It has 21 forks and 80 stars, so it feels like at least some folks are keeping an eye on it.

yarikoptic replied:
I think there is still remaining confusion somewhere.  We have two actions and they are complimentary:

- `codespell-project/codespell-problem-matcher@v1` -- this one is just to annotate in the github "files" view where typos are observed. handy but not mandatory, not as popular, no recent developments
- `codespell-project/actions-codespell@v2` -- the actual action to run `codespell` itself. 

so I would recommend to keep both. First first one makes no sense without the 2nd one

bolinfest replied:
I am opting to keep both. I am saying that instead of referencing the actions by tag (`@v1`, `@v2`), I would like to reference them by full commit hash so the behavior of the action cannot change out from under us.

yarikoptic replied:
ah, gotcha -- I just pushed replacing with commit hexshas

bolinfest reviewed (changes requested):
Out of an abundance of caution, can you please peg these to specific commits rather than tags? Since we don't own these CI jobs, they could change out from under us if we use tags.

bolinfest commented on .github/workflows/codespell.yml:
```diff
@@ -0,0 +1,25 @@
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4
```
This one is fine: this is GitHub's responsibility and if this does something malicious, the whole Internet will break anyway. So please keep this as `@v4` so it's easier to eyeball that we're doing something canonical.

yarikoptic committed:
[DATALAD RUNCMD] PIN github actions to their commits to prevent malicious code ran

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "pin-github-action .github/workflows/codespell.yml --allow 'actions/*'",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

bolinfest reviewed (changes requested):
sorry, one last thing

bolinfest commented:
Thanks @yarikoptic: I'm excited to see this in action!
